stdlib/private/OSLog/OSLogMessage.swift

```diff
@@ -0,0 +1,309 @@
+/// different types) that have to be serialized to a byte buffer. The arguments
+/// are captured within closures and stored in an array. The closures accept an
+/// instance of `OSLogByteBufferBuilder`, and when invoked, serialize the
+/// argument using the passed `OSLogByteBufferBuilder` instance.
+public struct OSLogArguments {
+  /// An array of closures that captures arguments of possibly different types.
+  internal var argumentClosures: [(inout OSLogByteBufferBuilder) -> ()]
+  /// Sum total of the byte size of the arguments that are tracked.
+  internal var byteCount: Int
+
+  public init() {
+    argumentClosures = []
+    byteCount = 0
+  }
+
+  /// Append a byte-sized header, constructed by
+  /// `OSLogMessage.appendInterpolation`, to the tracked array of closures.
+  public mutating func append(_ header: UInt8) {
+    argumentClosures.append({ $0.serialize(header) })
+    byteCount += OSLogByteBufferBuilder.numBytesForEncoding(UInt8.self)
+  }
+
+  /// Append an (autoclosured) interpolated expression of type Int, passed to
+  /// `OSLogMessage.appendInterpolation`, to the tracked array of closures.
+  public mutating func append(_ value: @escaping () -> Int) {
+    argumentClosures.append({ $0.serialize(value()) })
+    byteCount += OSLogByteBufferBuilder.numBytesForEncoding(Int.self)
+  }
+
+  internal func serialize(into bufferBuilder: inout OSLogByteBufferBuilder) {
+    argumentClosures.forEach { $0(&bufferBuilder) }
+  }
+}
+
+/// A struct that manages serialization of instances of specific types to a
+/// byte buffer. The byte buffer is provided as an argument to the initializer
+/// so that its lifetime can be managed by the caller.
+internal struct OSLogByteBufferBuilder {
+  internal var position: UnsafeMutablePointer<UInt8>
+
+  /// Initializer that accepts a pointer to a preexisting buffer.
+  /// - Parameter bufferStart: the starting pointer to a byte buffer
+  ///   that must contain the serialized bytes.
+  internal init(_ bufferStart: UnsafeMutablePointer<UInt8>) {
+    position = bufferStart
+  }
+
+  /// Serialize a UInt8 value at the buffer location pointed to by `position`.
+  internal mutating func serialize(_ value: UInt8) {
+    position[0] = value
+    position += 1
+  }
+
+  /// Serialize an Int at the buffer location pointed to by `position`.
+  internal mutating func serialize(_ value: Int) {
+    let byteCount = OSLogByteBufferBuilder.numBytesForEncoding(Int.self)
+    let dest = UnsafeMutableRawBufferPointer(start: position, count: byteCount)
+    withUnsafeBytes(of: value) { dest.copyMemory(from: $0) }
+    position += byteCount
+  }
+
+  /// Return the number of bytes needed for serializing an UInt8 value.
+  internal static func numBytesForEncoding(_ typ: UInt8.Type) -> Int {
+    return 1
+  }
+
+  /// Return the number of bytes needed for serializing an `Int` value.
+  internal static func numBytesForEncoding(_ typ: Int.Type) -> Int {
+    return Int.bitWidth / 8
```

Comment:
I guess Swift doesn't have to care about systems where a byte isn't 8 bits. (Clang does, though).